packages/git/src/browser/dirty-diff/dirty-diff-manager.ts

```diff
@@ -0,0 +1,263 @@
+        return this.preferences["git.editor.decorations.enabled"];
+    }
+
+    protected async handlePreferenceChange(event: PreferenceChangeEvent<GitConfiguration>): Promise<void> {
+        const { preferenceName, newValue } = event;
+        if (preferenceName === "git.editor.decorations.enabled") {
+            const enabled = !!newValue;
+            const models = await this.allModels();
+            for (const model of models) {
+                model.enabled = enabled;
+                model.update();
+            }
+        }
+    }
+
+    protected async allModels(): Promise<DirtyDiffModel[]> {
+        const models = [];
+        const uris = this.models.keys();
+        for (const uri of uris) {
+            const reference = await this.models.acquire(uri);
+            models.push(reference.object);
+            reference.dispose();
+        }
+        return models;
+    }
+
+    protected async getModel(uri: string): Promise<DirtyDiffModel | undefined> {
+        if (this.models.has(uri)) {
+            const reference = await this.models.acquire(uri);
+            const model = reference.object;
+            reference.dispose();
+            return model;
+        }
+        return undefined;
+    }
+
+}
+
+export interface DirtyDiffUpdate extends DirtyDiff {
+    readonly uri: string;
+}
+
+export class DirtyDiffModel implements Disposable {
+
+    enabled = true;
+
+    protected dirty = true;
+    protected staged: boolean;
+    protected previousContent: string[];
+    protected currentContent: string[];
+
+    protected readonly onDirtyDiffUpdateEmitter = new Emitter<DirtyDiffUpdate>();
+    readonly onDirtyDiffUpdate: Event<DirtyDiffUpdate> = this.onDirtyDiffUpdateEmitter.event;
+    protected readonly updateDelayer = new DirtyDiffModel.Throttler(200);
+
+    constructor(
+        readonly uri: string,
+        protected readonly readGitResource: DirtyDiffModel.GitResourceReader
+    ) { }
+
+    update(): void {
+        const enabled = this.enabled && this.dirty;
+        const previous = enabled ? this.previousContent : [];
+        const current = enabled ? this.currentContent : [];
+        if (!previous || !current) {
+            return;
+        }
+        const uri = this.uri;
+        this.updateDelayer.push(() => {
+            const dirtyDiff = enabled
+                ? DirtyDiffModel.computeDirtyDiff(previous, current)
+                : <DirtyDiff>{ added: [], removed: [], modified: [] };
+            const dirtyDiffUpdate = <DirtyDiffUpdate>{ uri, ...dirtyDiff };
+            this.onDirtyDiffUpdateEmitter.fire(dirtyDiffUpdate);
+        });
+    }
+
+    handleDocumentChanged(documentContent: string): void {
+        this.currentContent = DirtyDiffModel.splitLines(documentContent);
+        this.update();
+    }
+
+    async handleGitStatusUpdate(repository: Repository, relevantChanges: GitFileChange[]): Promise<void> {
+        const noRelevantChanges = relevantChanges.length < 0;
```

Comment:
This must be a typo or, a JS black magic, or I am just blind. `length` should be a non-negative integer, so it's always `false`.

Reply:
🙊 
fixed.